File: src/types/workflow.ts
import type { EffortLevel, PermissionMode } from "@/types"

/** What an agent node does — carries a built-in behavior; the edge supplies the
 *  content, so downstream nodes need no hand-written task. */
export type Intent = "plan" | "code" | "review" | "revise" | "custom"

export interface AgentTaskConfig {
  intent: Intent
  model: string
  effort: EffortLevel
  /** Feature description (plan/custom) or optional extra instructions. */
  prompt: string
  branchHint?: string | null
  /** Mode override — only honored for custom; other intents always run under
   *  their intent's mode. */
  permissionMode?: PermissionMode | null
}

/** A node's behavior, internally tagged by `type` to match serde. Gate carries
 *  no config — it's a pure user sign-off checkpoint between agent steps. */
export type NodeKind =
  | { type: "start" }
  | ({ type: "agentTask" } & AgentTaskConfig)
  | { type: "gate" }

export interface WorkflowNode {
  id: string
  label: string
  kind: NodeKind
  position?: { x: number; y: number } | null
}

export interface WorkflowEdge {
  id: string
  source: string
  target: string
}

export interface WorkflowGraph {
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
}

export interface Workflow {
  id: string
  projectId: string
  name: string
  graph: WorkflowGraph
  createdAt: string
  updatedAt: string
}

export type RunStatus =
  | "pending"
  | "running"
  | "paused"
  | "completed"
  | "failed"
  | "canceled"

export type NodeRunStatus =
  | "pending"
  | "running"
  | "done"
  | "failed"
  | "skipped"
  | "paused"
  | "awaitingInput"

export interface WorkflowRun {
  id: string
  workflowId?: string | null
  projectId: string
  groupId: string
  status: RunStatus
  error?: string | null
  createdAt: string
  updatedAt: string
}

export interface WorkflowNodeRun {
  runId: string
  nodeId: string
  status: NodeRunStatus
  sessionId?: string | null
  output?: string | null
  error?: string | null
}

export interface WorkflowRunView {
  run: WorkflowRun
  nodes: WorkflowNodeRun[]
}

File: src-tauri/src/domain/workflow.rs
//! Workflow graph domain types: a user-authored DAG of agent tasks (the
//! definition) and its executions (runs + per-node state). The graph is an
//! opaque JSON document the React Flow editor round-trips; the executor reads
//! `kind`/`edges` and ignores `position`.

use serde::{Deserialize, Serialize};

use crate::domain::{EffortLevel, PermissionMode, SessionRole};

/// A persisted workflow definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Workflow {
    pub id: String,
    pub project_id: String,
    pub name: String,
    pub graph: WorkflowGraph,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowGraph {
    #[serde(default)]
    pub nodes: Vec<WorkflowNode>,
    #[serde(default)]
    pub edges: Vec<WorkflowEdge>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowNode {
    pub id: String,
    #[serde(default)]
    pub label: String,
    pub kind: NodeKind,
    /// Canvas position — carried for the editor, ignored by the executor.
    #[serde(default)]
    pub position: Option<NodePosition>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct NodePosition {
    pub x: f64,
    pub y: f64,
}

/// The node's behavior. `HumanGate`/`Action` arrive in later phases.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum NodeKind {
    /// Topological root; synthesized if absent. Carries no agent.
    Start,
    /// Run an agent task (provider/model/mode/effort + prompt).
    AgentTask(AgentTaskConfig),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentTaskConfig {
    pub model: String,
    pub permission_mode: PermissionMode,
    pub effort: EffortLevel,
    pub role: SessionRole,
    pub prompt: String,
    /// Open the worktree on a named branch (e.g. `feat/x`); else `warden/<id>`.
    #[serde(default)]
    pub branch_hint: Option<String>,
    /// Whether this node edits code (needs the shared coding worktree).
    #[serde(default)]
    pub writes_code: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowEdge {
    pub id: String,
    pub source: String,
    pub target: String,
}

/// A single execution of a workflow.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowRun {
    pub id: String,
    pub workflow_id: Option<String>,
    pub project_id: String,
    pub group_id: String,
    pub status: RunStatus,
    pub error: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RunStatus {
    Pending,
    Running,
    Paused,
    Completed,
    Failed,
    Canceled,
}

impl RunStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            RunStatus::Pending => "pending",
            RunStatus::Running => "running",
            RunStatus::Paused => "paused",
            RunStatus::Completed => "completed",
            RunStatus::Failed => "failed",
            RunStatus::Canceled => "canceled",
        }
    }
    pub fn parse(s: &str) -> Option<Self> {
        Some(match s {
            "pending" => RunStatus::Pending,
            "running" => RunStatus::Running,
            "paused" => RunStatus::Paused,
            "completed" => RunStatus::Completed,
            "failed" => RunStatus::Failed,
            "canceled" => RunStatus::Canceled,
            _ => return None,
        })
    }
}

/// Per-node execution state, overlaying the graph (the node's session row stays
/// the source of truth for transcript/cost/status).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowNodeRun {
    pub run_id: String,
    pub node_id: String,
    pub status: NodeRunStatus,
    pub session_id: Option<String>,
    pub output: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum NodeRunStatus {
    Pending,
    Running,
    Done,
    Failed,
    Skipped,
}

impl NodeRunStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            NodeRunStatus::Pending => "pending",
            NodeRunStatus::Running => "running",
            NodeRunStatus::Done => "done",
            NodeRunStatus::Failed => "failed",
            NodeRunStatus::Skipped => "skipped",
        }
    }
    pub fn parse(s: &str) -> Option<Self> {
        Some(match s {
            "pending" => NodeRunStatus::Pending,
            "running" => NodeRunStatus::Running,
            "done" => NodeRunStatus::Done,
            "failed" => NodeRunStatus::Failed,
            "skipped" => NodeRunStatus::Skipped,
            _ => return None,
        })
    }
}
